be/src/storage/lake/local_pk_index_manager.cpp

```diff
@@ -0,0 +1,157 @@
+                }
+            }
+        }
+
+        auto debug_vector_info = [](std::vector<int64_t> vector) -> std::string {
+            std::string result;
+            for (int i = 0; i < vector.size(); i++) {
+                if (i != 0) {
+                    result.append(",");
+                }
+                result += std::to_string(vector[i]);
+            }
+            return result;
+        };
+
+        int64_t t_end = MonotonicMillis();
+        LOG(INFO) << "finish gc local persistent index dir: " << pk_path
+                  << ", found tablet not in the worker, tablet_ids: " << debug_vector_info(not_in_worker_tablet_ids)
+                  << ", data_dir changed tablet_ids: " << debug_vector_info(dir_changed_tablet_ids)
+                  << ", and removed dir successfully, tablet_ids: " << debug_vector_info(removed_dir_tablet_ids)
+                  << ", cost:" << t_end - t_start << "ms";
+    }
+}
+
+void LocalPkIndexManager::evict(UpdateManager* update_manager) {
+    for (DataDir* data_dir : StorageEngine::instance()->get_stores()) {
+        bool need_evict = false;
+        auto space_info_or = FileSystem::Default()->space(data_dir->path());
+        if (space_info_or.ok()) {
+            auto space_info = *space_info_or;
+            need_evict = (double)space_info.free < (double)space_info.capacity * config::starlet_cache_evict_low_water;
+        }
+        if (!need_evict) {
+            continue;
+        }
+
+        auto pk_path = data_dir->get_persistent_index_path();
+        LOG(INFO) << "start to evict local persistent index dir:" << pk_path;
+        int64_t t_start = MonotonicMillis();
+
+        std::set<std::string> tablet_ids;
+        Status ret = fs::list_dirs_files(pk_path, &tablet_ids, nullptr);
+        if (!ret.ok()) {
+            LOG(WARNING) << "fail to walk dir. path=[" + pk_path << "] error[" << ret.to_string() << "]";
+            continue;
+        }
+
+        std::vector<int64_t> tablet_ids_to_be_evicted;
+        std::vector<int64_t> removed_dir_tablet_ids;
+
+        for (const auto& tablet_id : tablet_ids) {
+            auto tablet_pk_path = pk_path + "/" + tablet_id;
+            int64_t id = std::stoll(tablet_id);
+            auto mtime_or = FileSystem::Default()->get_file_modified_time(tablet_pk_path);
+            auto mtime = *mtime_or;
+            auto now = std::chrono::duration_cast<std::chrono::seconds>(
+                               std::chrono::system_clock::now().time_since_epoch())
+                               .count();
+            if (now - mtime > config::lake_local_pk_index_unused_threshold) {
+                tablet_ids_to_be_evicted.emplace_back(id);
+                if (update_manager->try_remove_primary_index_cache(id)) {
+                    if (StorageEngine::instance()->clear_persistent_index(data_dir, id, tablet_pk_path).ok()) {
+                        removed_dir_tablet_ids.push_back(id);
+                    }
+                }
```

Comment:
No need to call `try_lock_pk_index_shard` like above?

Reply:
+1, lock should be hold to avoid apply thread is building persistent index when clear persistent index